### ISSUE TITLE
Rain(daily) and Rain(monthly) are Incorrect

### ISSUE THREAD

**Issue by mbking1** [BME-nodeservers/docs#1]
Rain(daily) is displaying the Rain rate instead of the daily accumulation and Rain(monthly) appears to be displaying the Rain(weekly) value.  I've looked at all other values except for some of the wind values and they look OK.  I'll review the wind values today.  Attached is the log file:

[AmbientWeather_1-24-2022_81823_AM.txt](https://github.com/BME-nodeservers/docs/files/7926196/AmbientWeather_1-24-2022_81823_AM.txt)
.

**Comment by bpaauwe** [comment]
The various rain values are determined by the field names in the data.  Without knowing what the data looks like, I can't determine if there's a problem with the mapping of field names to data types or not. 

The rain rate comes from a field name called "eventrain".  I don't see any place in the log where it is setting the rain rate.

Daily rain comes from field name "dailyrain", which should be accumulation.

**Comment by mbking1** [comment]
Hello Bob,

I was watching the data come in during a rain event this morning and compared the node server data to the Ambientweather.net info. This is how I saw the issues. 

“dailyrain” certainly seems like the right value to pull to get daily accumulation, but that wasn’t what was displayed in that node server attribute. One suggestion would be to compare simplextech’s mapping for AmbientWeather in Polyglot 2. That node server had the correct values. I know you guys worked to combine efforts into one for Polyglot 3. 

I could also contact Ambient Weather to see how they have it mapped in their iOS application. 

Did you see any issues with the Rain Monthly mapping?

Mark

**Comment by mbking1** [comment]
Bob,

I did a little digging and found the device parameters in the AmbientWeather API docs located at:

https://github.com/ambient-weather/api-docs/wiki/Device-Data-Specs <https://github.com/ambient-weather/api-docs/wiki/Device-Data-Specs>

dailyrainin = Daily Rain, in
hourlyrainin = Rain Rate, in/hr
weeklyrainin = Weekly Rain, in

Mark

**Comment by bpaauwe** [comment]
Yes, that's how it's mapped.  Same as what simplextech’s PG2 used.

**Comment by mbking1** [comment]
I’m at a lose then.  Here are screen shots from AmbientWeather.net <http://ambientweather.net/> and the node server and they don’t match Rain Daily.  During the rain event this morning, Rain (daily) in the node server was displaying the Rain Rate from AmbientWeather.net.  Not sure what to do next.





Mark

**Comment by mbking1** [comment]
Hello Bob,

I’m no expert in Python, but I found what appears to be some mapping issues in the code.  This is from the Polyglot 2 code,; I couldn’t find the Polyglot 3 code, so these mapping issue may not exist there.  The code didn’t match to what I would have expected and might be the problem I’m seeing.


The code includes this:

for node in self.nodes:
            if self.nodes[node].id == 'pressure':
                self.set_driver(node, 'GV0', d, 'baromrelin')
                if 'baromabsin' in d:
                    self.set_driver(node, 'ST', d, 'baromabsin')

                    trend = self.nodes[node].updateTrend(d['baromabsin'])
                    self.nodes[node].setDriver('GV1', trend, report = True, force = True)
            elif self.nodes[node].id == 'temperature':
                self.set_driver(node, 'ST', d, 'tempf')
                self.set_driver(node, 'GV0', d, 'feelsLike')
                self.set_driver(node, 'GV1', d, 'dewPoint')
                #self.set_driver(node, 'GV2', d, 'heatIndex')
                #self.set_driver(node, 'GV3', d, 'windchill')
            elif self.nodes[node].id == 'humidity':
                self.set_driver(node, 'ST', d, 'humidity')
            elif self.nodes[node].id == 'wind':
                self.set_driver(node, 'ST', d, 'windspeedmph')
                self.set_driver(node, 'GV0', d, 'winddir')
                self.set_driver(node, 'GV1', d, 'windgustmph')
                #self.set_driver(node, 'GV2', d, 'windgustdir')
            elif self.nodes[node].id == 'precipitation':
                #self.set_driver(node, 'ST', d, 'rainrate')
                self.set_driver(node, 'GV0', d, 'hourlyrainin')
                self.set_driver(node, 'GV1', d, 'dailyrainin')
                self.set_driver(node, 'GV2', d, 'weeklyrainin')
                self.set_driver(node, 'GV3', d, 'monthlyrainin')
                self.set_driver(node, 'GV4', d, 'yearlyrainin')
            elif self.nodes[node].id == 'light':
                self.set_driver(node, 'ST', d, 'uv')
                self.set_driver(node, 'GV0', d, 'solarradiation')
            elif self.nodes[node].id == 'indoor':
                self.set_driver(node, 'ST', d, 'tempinf')
                self.set_driver(node, 'GV0', d, 'humidityin'


And this:

class PrecipitationNode(polyinterface.Node):
    id = 'precipitation'
    drivers = [
            {'driver': 'ST', 'value': 0, 'uom': 24},  # rate
            {'driver': 'GV0', 'value': 0, 'uom': 105}, # hourly
            {'driver': 'GV1', 'value': 0, 'uom': 105}, # daily
            {'driver': 'GV2', 'value': 0, 'uom': 105}, # weekly
            {'driver': 'GV2', 'value': 0, 'uom': 105}, # monthly
            {'driver': 'GV2', 'value': 0, 'uom': 105}  # yearly
            ]

The “ST” variable is commented out in the first set of code, but active in the second.  Also, the “GV2” is duplicated in the second set of code.  Seems like they would match one for one.  Again, not an expert, but it seemed a possible answer to the issue I’m having.

Thanks,
Mark

**Comment by bpaauwe** [comment]
The screenshots didn't seem to get attached.

The PG3 code is very different from the either of the PG2 node servers.   The nodes are defined by the data found in the query since each station type is a little bit different.

Rain is always a pain to debug because it never seems to be raining when trying to debug it.   I don't have an Ambient Weather station so my testing is with a remote station which makes it even more difficult.   The code looks correct, but without seeing the actual raw data from ambientweather.net, I can't determine if what is displayed matches that or not.

**Comment by mbking1** [comment]
Bob,

I could send you my weather station info so you could monitor it with the node server and then set a time to put water in the rain sensor so you could look at the incoming data. As long as I’m not adding more than a few hundredths of an inch, it won’t throw off my numbers very much. 

Let me know if that would help or is there something else I could do. 

Mark
